README.md
# 介绍
******
环信小程序demo是基于环信im SDK开发的一款即时通讯的小程序。这个demo可以帮助开发者们更轻松的集成环信SDK。可扫码体验：
![huanxinIM](./src/images/huanxindemo.jpg)
demo 包含以下功能
  - 最近通话
  - 通讯录
  - 通知（加好友、加群）
  - 设置
# 在本地跑起来
拉取代码，导入开发者工具即可运行起来。

# 数据结构
```
登录页：
		login: {
			name:'',
			psd: '',
			grant_type: 'password',
		}
		
注册页：
		register: {
			username: '',
			password: ''
		}
		
通讯录页：
	   member:[],   //好友列表
	   
聊天页：
	   chatMsg:[{
			info:{
		        to:''         
			},
			username:'',      //用户名
			yourname:'',      //好友名
			msg: {
				type:'',    
				data:''
			},
			style:'',       //样式
			time:'',
			mid:''        //message ID
	   }]
	   
globalData: 
	userInfo: '',  //用户微信授权信息
	chatMsg: []   //用于存储离线消息
	unReadMessageNum: 0, //未读消息数
	saveFriendList: [],//加好友申请
	saveGroupInvitedList: [], //加群邀请
	isIPX: false //是否为iphone X //是否为iphoneX
	
缓存：
   myUsername: ''    //缓存登录用户名	   
   yourname + myName:''  //以用户名跟好友名为key来缓存聊天记录
```
# 项目结构
```shell
|- comps 自定义组件目录
    |- addfriend 添加好友页
    |-chat 聊天页面
    |-swipedelete 测滑删除
    |-toast toast
|-images demo中用到的图片 还有表情
|-pages 功能页面
    |-register 注册页
    |-login 登录页
    |-login_token token登录页
    |-chat 最近联系人页（通话）
        |-chatroom 聊天室页
    |-main 联系人页
        |-add_new 加好友页
        |-group 群组页
            |-groupSetting 群组设置页
    |-notificaton 通知页
        |-notificaton_friendDetail 加好友通知页
        |-notificaton_groupDetail 加群组通知页
    |-setting 设置页
        |-setting_general 通用设置页
|-utils 工具类和sdk的一些配置
|-sdk 环信sdk
|-app.js 小程序根实例，存放一些全局变量，注册监听事件
|-app.json 注册页面以及全局的一些配置
|-app.wxss 一些全局样式
|-project.config.json工程的一些配置，和开发者工具 “详情” 中的设置一样
```
# 可以复用的代码
如果想快速搭建起一个有im能力的小程序，可以选择复用demo中的代码，其中utils以帮助快速集成sdk，comps > chat是聊天页。当然其他部分需要复用的，如常见的测滑删除、联系人按字母分类、通知的订阅模式也可以去具体文件去找。

# 常见问题
+ 怎么发扩展消息？

  构造消息的时候msg.set(option)， option中传人ext字段，即可发送扩展消息,具体可以查看[文档](http://docs-im.easemob.com/im/applet/message#%E5%8F%91%E9%80%81%E6%96%87%E6%9C%AC%E6%B6%88%E6%81%AF_%E5%8D%95%E8%81%8A)。
+ 使用mpvue集成报错?

  .babelrc文件中修改 "plugins": ["transform-es2015-modules-commonjs"]，这里有一些开发者的[经验](https://blog.csdn.net/qq_43128835/article/details/95400888)。
+ 陌生人消息不上屏？

  目前demo并没有去实现，现在的逻辑是会话列表是从好友列表去查的，可以看pages/chat/chat.js getChatList()。
# 遇到的一些坑
- 聊天页面布局，input focus时页面滑动，光标会错位，官方的说法是，input focus时不能使用动画。
- 在scroll-view中无法触发onPullDownRefresh。
- margin-bottom 只有在下面还有元素时才生效。
- 使用enablePullDownRefresh看不见下拉动画，需要将backgroundTextStyle设为dark。
- 播放语音 onTimeUpdate在开发者工具上不能稳定触发。
- background-image只能用网络地址。
- 适配iphone X 使用wx.getSystemInfo() 返回的model在工具上是'iPhone X' 在真机上后面还会有其他字符，不要精确匹配。

坑还有很多，大家慢慢趟吧
# 写在最后
这期小程序demo只完善了单聊部分，后面还要优化群组的一些功能，关于图片消息缓存的问题，目前图片都是压缩过的，大约在200k左右，如果要做缓存自己可以使用storage去实现，但是要注意小程序只有10M的存储空间，还有一些功能demo没有去实现但是sdk是支持的，要用到的话大家可以去查[文档](https://webim.easemob.com/sdk/jsdoc/out/connection.html)

CHANGELOG.md
# 版本更新说明:
## ## v3.3.0 @ 2020-9-16
[IM SDK] 增加支持 promise
[IM SDK] 增加 onContactInvited、onContactDeleted、onContactAdded、onContactRefuse、onContactAgreed 好友相关的回调
[IM SDK] 增加 addContact、deleteContact、acceptInvitation、declineInvitation 代替原subscribe、removeRoster、subscribed、unsubscribed 好友操作 API
[IM SDK] 增加状态码 40，在 onError 中 type 为 40 会回调出因为socket断开导致发送失败的消息
[IM SDK] 修改默认的resource，以便区分web端和小程序端的用户
[IM SDK] 修改 getChatRooms 获取聊天室 API，去掉 apiUrl 参数
[IM SDK] 修改 构造 cmd 消息 API, 去掉 msg 参数
[IM SDK] 优化构造消API，使用 chatType 来区分消息类型（单聊/群聊/聊天室）
[IM SDK] 修复发送位置消息成功后并不执行 success 回调
[IM SDK] 增加容错处理
[demo] 修复小程序断网时发送单聊消息，未有发送成功和发送失败的区分。

## v3.2.2 @ 2020-8-25
[IM SDK] 创建群组增加是否同意参数参数
[IM SDK] 修复头条小程序消息延迟
[IM SDK] onError回调增加 error message

## v3.1.6 @ 2020-07-03
[音视频 SDK] 增加joinRoom api
[demo] 部分机型在视频会议时被来电打断后，无法继续推拉流，此时自动退出会议。


## v3.1.4 @ 2020-06-11
[IM SDK] 支持附件下载重定向
[IM SDK] 支持图片检测违规抛出单独的异常
[IM SDK] 增加分页获取聊天室成员api
[IM SDK] 修复收消息有延迟情况
[音视频 SDK] 增加断网重连
[demo] 用户删除小程序用户，小程序未实时更新，必须切换下页面联系人才会更新
[demo] 微信版本小程序，联系人界面下拉界面后，切换其他界面无效
[demo] 其他用户邀请小程序加入群组中，其他用户在群组中发消息，小程序聊天列表中未显示消息
[demo] 增加对不合法图片的提示


## v3.1.2 @ 2020-05-14
### Feature
+ [IM SDK] 增加上传修改群/聊天室公告、获取群/聊天室公告、上传/下载/删除群/聊天室文件、获取群/聊天室文件列表, 增加上传修改群/聊天室公告、获取群/聊天室公告、上传/下载/删除群/聊天室文件、获取群/聊天室文件列表
+ [IM SDK] 修改重连间隔
+ [IM SDK] 去掉对上传文件大小的限制, 由服务端来限制
+ [IM SDK] 修复拉历史消息bug
+ [IM SDK] 修复自定义消息没有time
+ [demo] 增加接受群组邀请功能，可以接收移动端的加群邀请
+ [demo] 同意好友申请后还会反向订阅
+ [demo] 用cover-view代替view 解决有时input内的文字会在toast之上


## v3.1.1 @ 2020-04-28
### Feature
+ [demo] 适配v3.1.1 IM SDK
+ [IM SDK] 更新私有协议, 与web端统一
+ [IM SDK] 增加漫游消息api
+ [IM SDK] 增加聊天室禁言、解除禁言、获取禁言列表、加入黑名单、移除黑名单、获取黑明单列表等api
+ [IM SDK] 增加聊天室、群组一键禁言、白名单等api
+ [IM SDK] 增加发送自定义消息
+ [IM SDK] 增加群组回执
+ [音视频 SDK] 兼容 v3.1.1 IM SDK
+ [音视频 SDK] 增加关闭摄像头的回调事件
+ [音视频 SDK] 修复不能销毁会议

demo从 IM SDK 2.0 升级 3.0涉及改动：
1、去掉 setPrensense 方法
2、查询群组更换api为 getGroup
3、Jid 与2.0不同，变为对象，需要自己拼成字符串
4、listgroupmemeber、getGroupInfo 返回 res.data 不是res.data.data
5、leaveGroupBySelf退出群、删除好友api变化
6、addGroupMembers邀请入群变为inviteToGroup
7、注册utils.registerUser变为conn.registerUser
8、xmppURL以及小程序后台socket域名变为wss://im-api-wechat.easemob.com/websocket

=====3.0=======

## v1.0.1 @ 2017-07-17

### Feature

* [sdk] 新增加入聊天室接口(joinChatRoom)
* [sdk] 新增成员加入聊天室的回调(memberJoinChatRoomSuccess)
* [demo] 新增发送位置消息功能
* [demo] 新增浏览收到的小视频功能

## v1.1.0 @ 2019-03-22
### Feature
+ [sdk] [demo] 增加token登录
+ [demo] 新版demo，修改ui
+ [demo] 增加搜索功能
+ [demo] 增加联系人按字母排序
+ [demo] 增加最近聊天按时间排序
+ [demo] 增加群组消息提醒
+ [demo] 增加测滑删除功能
+ [demo] 增加聊天历史分页
+ [demo] 增加用户名不区分大小写
+ [demo] 增加接收文件消息提示
+ [demo] 适配iphonex，以及 xs max等机型
+ [demo] 主页面由联系人页改为聊天页
+ [bug] 修改A给好友B发语音消息，B没有显示语音的未读消息数
+ [bug] iOS聊天界点击输入框进行输入时历史消息展示不合理
+ [bug] iOS端小程序收到消息时，会话界面来消息的提醒有时会没有提醒，只显示消息数
+ [bug] 语音消息时长为0

## v1.1.1 @ 2019-04-10
### Feature
+ [sdk] 增加重连机制
+ [demo] 实时更新联系人列表
+ [bug] 修复部分已知bug

## v1.2.0 @ 2019-06-22
### Feature
+ 增加消息状态，比如断网时发的消息显示失败。
+ 增加socket连接成功的提示
+ 修改了语音消息播放时再下载
+ 修复聊天页面切后台，再切前台收到的离线消息有重复
+ 由rest1迁移到rest2后开始校验token,导致附件消息收不到
+ 语音发送成功后点击听取后，语音依然闪烁动画
+ 联系人分类为#，显示问题。
+ sdk增加onSocketConnected事件 – socket连接成功
+ sdk onError 增加type='sendMsgError'发送消息失败
+ sdk 重连时关闭上次的socket

## v1.3.0 @ 2020-02-10
### Feature
+ 增加音视频会议功能


src/app.js

// require("sdk/libs/strophe");
let WebIM = wx.WebIM = require("utils/WebIM")["default"];
let msgStorage = require("comps/chat/msgstorage");
let msgType = require("comps/chat/msgtype");
let ToastPannel = require("./comps/toast/toast");
let disp = require("utils/broadcast");
let logout = false;

//emedia_for_miniProgram-test是沙箱环境测试版，线上环境请用emedia_for_miniProgram这个文件
const emedia = wx.emedia = require("./emedia/emedia_for_miniProgram") 

console.log('emedia', emedia)
console.log('WebIM', WebIM)

function ack(receiveMsg){
	// 处理未读消息回执
	var bodyId = receiveMsg.id;         // 需要发送已读回执的消息id
	var ackMsg = new WebIM.message("read", WebIM.conn.getUniqueId());
	ackMsg.set({
		id: bodyId,
		to: receiveMsg.from
	});
	WebIM.conn.send(ackMsg.body);
}

function onMessageError(err){
	if(err.type === "error"){
		wx.showToast({
			title: err.errorText
		});
		return false;
	}
	return true;
}

function getCurrentRoute(){
	let pages = getCurrentPages();
	if (pages.length > 0) {
		let currentPage = pages[pages.length - 1];
		return currentPage.route;
	}
	return '/'
}

// 不包含陌生人版本
// function calcUnReadSpot(message){
// 	let myName = wx.getStorageSync("myUsername");
// 	let members = wx.getStorageSync("member") || []; //好友
// 	var listGroups = wx.getStorageSync('listGroup')|| []; //群组
// 	let allMembers = members.concat(listGroups)
// 	let count = allMembers.reduce(function(result, curMember, idx){
// 		let chatMsgs;
// 		if (curMember.groupid) {
// 			chatMsgs = wx.getStorageSync(curMember.groupid + myName.toLowerCase()) || [];
// 		}else{
// 			chatMsgs = wx.getStorageSync(curMember.name.toLowerCase() + myName.toLowerCase()) || [];
// 		}
// 		return result + chatMsgs.length;
// 	}, 0);
// 	getApp().globalData.unReadMessageNum = count;
// 	disp.fire("em.xmpp.unreadspot", message);
// }
// 包含陌生人版本
function calcUnReadSpot(message){
	let myName = wx.getStorageSync("myUsername");
	wx.getStorageInfo({
		success: function(res){
			let storageKeys = res.keys
			let newChatMsgKeys = [];
			let historyChatMsgKeys = [];
			storageKeys.forEach((item) => {
				if (item.indexOf(myName) > -1 && item.indexOf('rendered_') == -1) {
					newChatMsgKeys.push(item)
				}
			})
			let count = newChatMsgKeys.reduce(function(result, curMember, idx){
				let chatMsgs;
				chatMsgs = wx.getStorageSync(curMember) || [];
				return result + chatMsgs.length;
			}, 0)

			getApp().globalData.unReadMessageNum = count;
			disp.fire("em.xmpp.unreadspot", message);
		}
	})
}

function saveGroups(){
	var me = this;
	return WebIM.conn.getGroup({
		limit: 50,
		success: function(res){
			wx.setStorage({
				key: "listGroup",
				data: res.data
			});
		},
		error: function(err){
			console.log(err)
		}
	});
}

App({
	ToastPannel,
	globalData: {
		unReadMessageNum: 0,
		userInfo: null,
		saveFriendList: [],
		saveGroupInvitedList: [],
		isIPX: false //是否为iphone X
	},

	conn: {
		closed: false,
		curOpenOpt: {},
		open(opt){
			wx.showLoading({
			  	title: '正在初始化客户端...',
			  	mask: true
			})
			this.curOpenOpt = opt;
			WebIM.conn.open(opt);
			this.closed = false;
		},
		reopen(){
			if(this.closed){
				//this.open(this.curOpenOpt);
				WebIM.conn.open(this.curOpenOpt);
				this.closed = false;
			}
		}
	},

	// getPage(pageName){
	// 	var pages = getCurrentPages();
	// 	return pages.find(function(page){
	// 		return page.__route__ == pageName;
	// 	});
	// },

	onLaunch(){
		// 调用 API 从本地缓存中获取数据
		wx.setInnerAudioOption({obeyMuteSwitch: false})
		var me = this;
		var logs = wx.getStorageSync("logs") || [];
		logs.unshift(Date.now());
		wx.setStorageSync("logs", logs);

		// 
		disp.on("em.main.ready", function(){
			calcUnReadSpot();
		});
		disp.on("em.chatroom.leave", function(){
			calcUnReadSpot();
		});
		disp.on("em.chat.session.remove", function(){
			calcUnReadSpot();
		});
		disp.on('em.chat.audio.fileLoaded', function(){
			calcUnReadSpot()
		});

		disp.on('em.main.deleteFriend', function(){
			calcUnReadSpot()
		});
		disp.on('em.chat.audio.fileLoaded', function(){
			calcUnReadSpot()
		});
		
		// 
		WebIM.conn.listen({
			onOpened(message){
				console.log('im登录成功')
				// WebIM.conn.setPresence();
				if(getCurrentRoute() == "pages/login/login" || getCurrentRoute() == "pages/login_token/login_token"){
					me.onLoginSuccess(wx.getStorageSync("myUsername").toLowerCase());
				}

				let identityToken = WebIM.conn.context.accessToken
				let identityName = WebIM.conn.context.jid

				// service.setup({"tktId":"13H9ZH0001TE29IFLK000C7TK2","url":"wss://172.17.2.55/confr/multipeople?CONFRID=13H9ZH0001TE29IFLK000C1M3&forward=127.0.0.1&port=9092","confrId":"13H9ZH0001TE29IFLK000C1M3","password":"","type":"communication_mix","memName":"easemob-demo#chatdemoui_zdtest@easemob.com","hmac":"SSZpG1K0U6cuIl8TWWV06yXgBaQ=","timestamp":1575962109914,"rights":15}
				// )
				// service.join()
				//"{"tktId":"13H851UX8XTE10GSUIQ00C1TK2","url":"wss://rtc-turn4-hsb.easemob.com/ws?CONFRID=13H851UX8XTE10GSUIQ00C1&forward=10.29.117.29&port=9092","confrId":"13H851UX8XTE10GSUIQ00C1","password":"","type":"communication_mix","memName":"easemob-demo#chatdemoui_zdtest@easemob.com","hmac":"Ak0tITkTxIp+RFlDC/B3LKm2iMw=","timestamp":1575873756927,"rights":7}"
				// emedia.mgr.createConfr({
				// 	identityName: 'easemob-demo#chatdemoui_zdtest4@easemob.com',
				//  	identityToken: identityToken,//'YWMtLFeEbBpOEeqD-sMgAnWU5U1-S6DcShHjkNXh_7qs2vUy04pwHuER6YGUI5WOSRNCAwMAAAFu6V9A4ABPGgDCHHYPZf0jtQbrjH97smaj5nqfv0jQI3WQ2Idfa30bqg',
				//  	confrType: 11,
				// 	password: '',
				// 	success: function(data){
				// 		var ticket = JSON.parse(data.ticket)
				// 		//ticket.url = ticket.url//.replace('localhost', '172.17.2.55')
				// 		var ssss = service.setup(ticket)
				// 		console.log('ssss', ssss)
				// 		service.join()

				// 		wx.emedia.onAddStream=function(data){
				// 			console.log('onAddStream', data)
				// 			getApp().globalData.subUrl = data.rtmp
				// 		}
				// 	}
				// })

			},
			onReconnect(){
				wx.showToast({
					title: "重连中...",
					duration: 2000
				});
			},
			onSocketConnected(){
				wx.showToast({
					title: "socket连接成功",
					duration: 2000
				});
			},
			onClosed(){
				// wx.showToast({
				// 	title: "网络已断开",
				// 	icon: 'none',
				// 	duration: 2000
				// });
				wx.redirectTo({
						url: "../login/login"
					});
				me.conn.closed = true;
				WebIM.conn.close();
			},
			onInviteMessage(message){
				me.globalData.saveGroupInvitedList.push(message);
				disp.fire("em.xmpp.invite.joingroup", message);
				// wx.showModal({
				// 	title: message.from + " 已邀你入群 " + message.roomid,
				// 	success(){
				// 		disp.fire("em.xmpp.invite.joingroup", message);
				// 	},
				// 	error(){
				// 		disp.fire("em.xmpp.invite.joingroup", message);
				// 	}
				// });
			},
			onReadMessage(message){
				//console.log('已读', message)
			},
			onPresence(message){
				console.log("onPresence", message);
				switch(message.type){
				case "unsubscribe":
					console.log('unsubscribe')
					// pages[0].moveFriend(message);
					break;
				// 好友邀请列表
				case "subscribe":
					if(message.status === "[resp:true]"){

					}
					else{
						// pages[0].handleFriendMsg(message);
						for (let i = 0; i < me.globalData.saveFriendList.length; i ++) {
					      	if(me.globalData.saveFriendList[i].from === message.from){
					      		me.globalData.saveFriendList[i] = message
					      		disp.fire("em.xmpp.subscribe");
					      		return;
					 		}
					    }
						me.globalData.saveFriendList.push(message);
						disp.fire("em.xmpp.subscribe");
					}
					break;
				case "subscribed":
					wx.showToast({
						title: "添加成功",
						duration: 1000
					});
					disp.fire("em.xmpp.subscribed");
					break;
				case "unsubscribed":
					// 延时1.5秒， 防止刚登录时和登录的toast重合
					setTimeout(() => {
						wx.showToast({
							title: message.from + "已退订",
							duration: 2000
						});
					}, 1500)
					
					disp.fire("em.xmpp.unsubscribed");
					break;
				case "direct_joined":
					saveGroups();
					wx.showToast({
						title: "已进群",
						duration: 1000
					});
					break;
				case "memberJoinPublicGroupSuccess":
					saveGroups();
					wx.showToast({
						title: "已进群",
						duration: 1000
					});
					break;
				case 'invite':
					let info = message.from + '邀请你加入群组'
					wx.showModal({
					  title: '提示',
					  content: info,
					  success (res) {
					    if (res.confirm) {
					      console.log('用户点击确定')
					      WebIM.conn.agreeInviteIntoGroup({
					      	invitee: WebIM.conn.context.userId,
							groupId: message.gid,
							success: () => {
								saveGroups();
								console.log('加入成功')
							}
					      })
					    } else if (res.cancel) {
					      console.log('用户点击取消')
					      WebIM.conn.rejectInviteIntoGroup({
					      	invitee: WebIM.conn.context.userId,
							groupId: message.gid
					      })
					    }
					  }
					})
					break;
				// 好友列表
				// case "subscribed":
				// 	let newFriendList = [];
				// 	for(let i = 0; i < me.globalData.saveFriendList.length; i++){
				// 		if(me.globalData.saveFriendList[i].from != message.from){
				// 			newFriendList.push(me.globalData.saveFriendList[i]);
				// 		}
				// 	}
				// 	me.globalData.saveFriendList = newFriendList;
				// 	break;
				// 删除好友
				case "unavailable":
					disp.fire("em.xmpp.contacts.remove");
					disp.fire("em.xmpp.group.leaveGroup", message);
					break;

				case 'deleteGroupChat':
					disp.fire("em.xmpp.invite.deleteGroup", message);
					break;

				case "leaveGroup": 
					disp.fire("em.xmpp.group.leaveGroup", message);
					break;

				case "removedFromGroup": 
					disp.fire("em.xmpp.group.leaveGroup", message);
					break;
				// case "joinChatRoomSuccess":
				// 	wx.showToast({
				// 		title: "JoinChatRoomSuccess",
				// 	});
				// 	break;
				// case "memberJoinChatRoomSuccess":
				// 	wx.showToast({
				// 		title: "memberJoinChatRoomSuccess",
				// 	});
				// 	break;
				// case "memberLeaveChatRoomSuccess":
				// 	wx.showToast({
				// 		title: "leaveChatRoomSuccess",
				// 	});
				// 	break;

				default:
					break;
				}
			},

			onRoster(message){
				// let pages = getCurrentPages();
				// if(pages[0]){
				// 	pages[0].onShow();
				// }
			},

			onVideoMessage(message){
				console.log("onVideoMessage: ", message);
				if(message){
					msgStorage.saveReceiveMsg(message, msgType.VIDEO);
				}
				calcUnReadSpot(message);
				ack(message);
			},

			onAudioMessage(message){
				console.log("onAudioMessage", message);
				if(message){
					if(onMessageError(message)){
						msgStorage.saveReceiveMsg(message, msgType.AUDIO);
					}
					calcUnReadSpot(message);
					ack(message);
				}
			},
			
			onCmdMessage(message){
				console.log("onCmdMessage", message);
				if(message){
					if(onMessageError(message)){
						msgStorage.saveReceiveMsg(message, msgType.CMD);
					}
					calcUnReadSpot(message);
					ack(message);
				}
			},

			// onLocationMessage(message){
			// 	console.log("Location message: ", message);
			// 	if(message){
			// 		msgStorage.saveReceiveMsg(message, msgType.LOCATION);
			// 	}
			// },

			onTextMessage(message){
				console.log("onTextMessage", message);
				if(message){
					if(onMessageError(message)){
						msgStorage.saveReceiveMsg(message, msgType.TEXT);
					}
					calcUnReadSpot(message);
					ack(message);

					if(message.ext.msg_extension){
						let msgExtension = JSON.parse(message.ext.msg_extension)
						let conferenceId = message.ext.conferenceId
						let password = message.ext.password
						disp.fire("em.xmpp.videoCall", {
							msgExtension: msgExtension,
							conferenceId: conferenceId,
							password: password
						});
					}
				}
			},

			onEmojiMessage(message){
				console.log("onEmojiMessage", message);
				if(message){
					if(onMessageError(message)){
						msgStorage.saveReceiveMsg(message, msgType.EMOJI);
					}
					calcUnReadSpot(message);
					ack(message);
				}
			},

			onPictureMessage(message){
				console.log("onPictureMessage", message);
				if(message){
					if(onMessageError(message)){
						msgStorage.saveReceiveMsg(message, msgType.IMAGE);
					}
					calcUnReadSpot(message);
					ack(message);
				}
			},

			onFileMessage(message){
				console.log('onFileMessage', message);
				if (message) {
					if(onMessageError(message)){
						msgStorage.saveReceiveMsg(message, msgType.FILE);
					}
					calcUnReadSpot(message);
					ack(message);
				}
			},

			// 各种异常
			onError(error){
				console.log('error', error)

				if (error.type == 40) { //send msg fail
					disp.fire("em.xmpp.error.sendMsgErr", error.failMsgs);
				}

				// 16: server-side close the websocket connection
				if(error.type == WebIM.statusCode.WEBIM_CONNCTION_DISCONNECTED && !logout){
					if(WebIM.conn.autoReconnectNumTotal >= WebIM.conn.autoReconnectNumMax){
						wx.showToast({
							title: "server-side close the websocket connection",
							duration: 1000
						});
						WebIM.conn.close();
						wx.redirectTo({
							url: "../login/login"
						});
						logout = true
					}
					return
				}
				// 8: offline by multi login
				if(error.type == WebIM.statusCode.WEBIM_CONNCTION_SERVER_ERROR){
					wx.showToast({
						title: "offline by multi login",
						duration: 1000
					});
					wx.redirectTo({
						url: "../login/login"
					});
				}
				if(error.type ==  WebIM.statusCode.WEBIM_CONNCTION_OPEN_ERROR){
					wx.hideLoading()
					disp.fire("em.xmpp.error.passwordErr");
					// wx.showModal({
					// 	title: "用户名或密码错误",
					// 	confirmText: "OK",
					// 	showCancel: false
					// });
				}
				if (error.type == WebIM.statusCode.WEBIM_CONNCTION_AUTH_ERROR) {
					wx.hideLoading()
					disp.fire("em.xmpp.error.tokenErr");
				}
				if (error.type == 16) {///sendMsgError
					// https://developers.weixin.qq.com/community/develop/doc/00084a400202787b54f8c9e6357800
					// 因为上面的原因 这里不要一直提示了
					return 
					console.log('socket_errorsocket_error', error)
					wx.showToast({
						title: "网络已断开",
						icon: 'none',
						duration: 2000
					});
					disp.fire("em.xmpp.error.sendMsgErr", error);
				}
			},
		});
		this.checkIsIPhoneX();
		
	},
	onShow(){
		// 从搜索页面进的时候退出后再回来会回到首页，此时并没有调用退出，导致登录不上
		// 判断当前是登录状态直接跳转到chat页面
		const pages = getCurrentPages();
		const currentPage = pages[pages.length - 1];
		// 选择图片或者拍照也会触发onShow，所以忽略聊天页面
		if (WebIM.conn.isOpened() && currentPage.route != "pages/chatroom/chatroom" && currentPage.route != "pages/groupChatRoom/groupChatRoom") {
			let myName = wx.getStorageSync("myUsername");
			wx.redirectTo({
				url: "../chat/chat?myName=" + myName
			});
		}
	},

	// onHide(){

	// 	WebIM.conn.close();
	// 	WebIM.conn.stopHeartBeat();
	// },

	// onUnload(){
	// 	WebIM.conn.close();
	// 	WebIM.conn.stopHeartBeat();
	// 	wx.redirectTo({
	// 		url: "../login/login?myName=" + myName
	// 	});
	// },

	onLoginSuccess: function(myName){
		wx.hideLoading()

		wx.redirectTo({
			url: "../chat/chat?myName=" + myName
		});

		// wx.redirectTo({
		// 	url: "../emediaInvite/emediaInvite?myName=" + myName
		// });
	},

	getUserInfo(cb){
		var me = this;
		if(this.globalData.userInfo){
			typeof cb == "function" && cb(this.globalData.userInfo);
		}
		else{
			// 调用登录接口
			wx.login({
				success(){
					wx.getUserInfo({
						success(res){
							me.globalData.userInfo = res.userInfo;
							typeof cb == "function" && cb(me.globalData.userInfo);
						}
					});
				}
			});
		}
	},
	checkIsIPhoneX: function() {
	    const me = this
	    wx.getSystemInfo({
	      	success: function (res) {
		        // 根据 model 进行判断
		        if (res.model.search('iPhone X') != -1) {
		          	me.globalData.isIPX = true
		        }
	      	}
	    })
	},

});

